### ISSUE TITLE
resource_github_team_repository cannot handle custom role after creation

### ISSUE THREAD

**Issue by joshua-hancox** [integrations/terraform-provider-github#1183]
## Affected Resource(s)

- resource_github_team_repository

## Expected Behavior
After creating a resource_github_team_repository with a custom role set in the permission field, making no changes, and re-running the plan, I expect no changes to be planned by terraform.

## Actual Behavior
After creating the resource and re-running the plan, terraform plan tries to change from the built in role that the custom role was based on, to the role that was originally set.

## Steps to Reproduce
e.g. config
```hcl
resource "github_team_repository" "custom_role" {
  team_id    = github_team.this.id
  repository = github_repository.this.name
  permission = "my-custom-role"
}
```
(assume that `my-custom-role` was created in github, and based on the built in maintainer role, with some added permissions)

1. Run terraform apply
2. Once the terraform apply has been applied, run terraform plan

The plan output will show:

```bash
  # github_team_repository.custom_role will be updated in-place
  ~ resource "github_team_repository" "custom_role" {
        id         = "5817097:my-repo-name"
      ~ permission = "maintainer" -> "my-custom-role"
    }
```

## Suggested fix

I think this issue comes from [this function](https://github.com/integrations/terraform-provider-github/blob/366aac0e7ff3a9c6f44c26c397ca4aefb1ec0803/github/util_permissions.go#L20) being called [here](https://github.com/integrations/terraform-provider-github/blob/main/github/resource_github_team_repository.go#L135) to guess what the permission is that is applied.

It assumes that there are only built in roles, which is no longer the case.

I have extended the Repository struct in go-github with [this PR](https://github.com/google/go-github/pull/2379) to include the role_name field in the repository that is returned(automatically includes the `.GetRoleName()` function), and also raised a [PR in this repo](https://github.com/integrations/terraform-provider-github/pull/1182) to bump to the latest version of that module.

Once that is complete I think we should be able to remove the function `getRepoPermission` altogether and amend `resource_github_team_repository` to do it's permission look up from:

```golang
	permName, permErr := getRepoPermission(repo.GetPermissions())
	if permErr != nil {
		return permErr
	}

	d.Set("permission", permName)
```
to
```hcl
	d.Set("permission", repo.GetRoleName())
```

If that makes sense, and we can merge the PR to bump the module then let me know and I can raise a new PR for the fix.

**Comment by joshuahancox** [comment]
Hey @kfcampbell - sorry to pester you, I noticed this behaviour when using the custom roles that we merged changes for recently. I wonder if you could take a look.

**Comment by kfcampbell** [comment]
@joshuahancox No worries at all. I'll set aside some time to troubleshoot this next week.

**Comment by joshua-hancox** [comment]
Hi @kfcampbell!

Could we merge https://github.com/integrations/terraform-provider-github/pull/1205/files so that I can continue working on this?

**Comment by joshua-hancox** [comment]
@kfcampbell did you have any time to look at this one?

**Comment by SharpEdgeMarshall** [comment]
This is blocking us from adopting terraform github, any news?

**Comment by kfcampbell** [comment]
I've merged the dependency PR and updated @joshua-hancox's PR with main before merging. 

Thank you for the contribution, @joshua-hancox!

**Comment by SharpEdgeMarshall** [comment]
@kfcampbell may we reopen this?

**Comment by joshua-hancox** [comment]
Apologies for the oversight @kfcampbell! I've raised a new PR that may help, reusing the existing function `getInvitationPermission()` that already does some conversion for this issue. 

Perhaps we should also rename the function? (since it is now not just related to invitations)

Could someone take a look, I am not 100% sure on what the edge case was that we broke before.

PR: https://github.com/integrations/terraform-provider-github/pull/1230
